README.md
# PHP URL library

[![Latest Stable Version](https://poser.pugx.org/josantonius/url/v/stable)](https://packagist.org/packages/josantonius/url)
[![License](https://poser.pugx.org/josantonius/url/license)](LICENSE)
[![Total Downloads](https://poser.pugx.org/josantonius/url/downloads)](https://packagist.org/packages/josantonius/url)
[![CI](https://github.com/josantonius/php-url/actions/workflows/ci.yml/badge.svg?branch=main)](https://github.com/josantonius/php-url/actions/workflows/ci.yml)
[![CodeCov](https://codecov.io/gh/josantonius/php-url/branch/main/graph/badge.svg)](https://codecov.io/gh/josantonius/php-url)
[![PSR1](https://img.shields.io/badge/PSR-1-f57046.svg)](https://www.php-fig.org/psr/psr-1/)
[![PSR4](https://img.shields.io/badge/PSR-4-9b59b6.svg)](https://www.php-fig.org/psr/psr-4/)
[![PSR12](https://img.shields.io/badge/PSR-12-1abc9c.svg)](https://www.php-fig.org/psr/psr-12/)

**Translations**: [Español](.github/lang/es-ES/README.md)

PHP library to access URL information.

Provides an improved replacement for the access to the components of a URL offered by PHP's
`parse_url` and `pathinfo` functions.

This library does not format the provided URL, it only makes it easier to access the components.
For something more advanced you can use something like `league/uri-components`.

---

- [Requirements](#requirements)
- [Installation](#installation)
- [Available Classes](#available-classes)
  - [Url Class](#url-class)
- [Usage](#usage)
- [Tests](#tests)
- [TODO](#todo)
- [Changelog](#changelog)
- [Contribution](#contribution)
- [Sponsor](#sponsor)
- [License](#license)

---

## Requirements

- Operating System: Linux | Windows.

- PHP versions: 8.1 | 8.2.

## Installation

The preferred way to install this extension is through [Composer](http://getcomposer.org/download/).

To install **PHP URL library**, simply:

```console
composer require josantonius/url
```

The previous command will only install the necessary files,
if you prefer to **download the entire source code** you can use:

```console
composer require josantonius/url --prefer-source
```

You can also **clone the complete repository** with Git:

```console
git clone https://github.com/josantonius/php-url.git
```

## Available Classes

### Url Class

`Josantonius\Url\Url`

Create a new instance:

```php
/**
 * If no URL is provided, the URL of the current page will be generated.
 * 
 * The generated URL will exclude ports 80 and 443 and include the rest.
 */
public function __construct(null|string $url = null);
```

Gets authority:

```php
/**
 * The authority, in "[user-info@][host][:port]" format.
 *
 * @var string URL authority or empty string.
 */
public readonly string $authority;
```

Gets the base URL:

```php
/**
 * The base URL, in "[scheme:][//domain][:port]" format.
 *
 * @var string Base URL or empty string.
 */
public readonly string $base;
```

Gets the path basename:

```php
/**
 * The path basename, in "[filename][.extension]" format.
 *
 * @var string URL path basename or empty string.
 */
public readonly string $basename;
```

Gets the path dirname:

```php
/**
 * The path dirname, in "[dirname]" format.
 *
 * @var string URL path dirname or empty string.
 */
public readonly string $dirname;
```

Gets the path basename extension:

```php
/**
 * The path basename extension, in "[extension]" format.
 *
 * @var string URL path basename extension or empty string.
 */
public readonly string $extension;
```

Gets the path filename:

```php
/**
 * The path filename, in "[filename]" format.
 *
 * @var string URL path filename or empty string.
 */
public readonly string $filename;
```

Gets fragment:

```php
/**
 * URL fragment in "[fragment]" format.
 *
 * @var string URL fragment or empty string.
 */
public readonly string $fragment;
```

Gets the full URL:

```php
public readonly string $full;
```

Gets hashed fragment:

```php
/**
 * URL hashed fragment in "[#fragment]" format.
 *
 * @var string URL hashed fragment or empty string.
 */
public readonly string $hash;
```

Gets host:

```php
/**
 * URL host in "[subdomain.][domain][.tld]" format.
 *
 * @var string URL host or empty string.
 */
public readonly string $host;
```

Gets path:

```php
/**
 * URL path in "[path]" format.
 *
 * @var string URL path or empty string.
 */
public readonly string $path;
```

Gets the query parameters:

```php
/**
 * URL query parameters in array format.
 *
 * @var array<string, mixed> URL query parameters or empty string.
 */
public readonly array $parameters;
```

Gets password:

```php
/**
 * URL password in "[password]" format.
 *
 * @var string URL password or empty string.
 */
public readonly string $password;
```

Gets port:

```php
/**
 * URL port in "[port]" format.
 *
 * @var string URL port or empty string.
 */
public readonly int|string $port;
```

Gets scheme:

```php
/**
 * URL scheme in "[scheme]" format.
 *
 * @var string URL scheme or empty string.
 */
public readonly string $scheme;
```

Gets path segments:

```php
/**
 * URL path segments in array format.
 *
 * @var string[] URL path segments or empty string.
 */
public readonly array $segments;
```

Gets query:

```php
/**
 * URL query in "[query]" format.
 *
 * @var string URL query or empty string.
 */
public readonly string $query;
```

Gets username:

```php
/**
 * URL username in "[username]" format.
 *
 * @var string URL username or empty string.
 */
public readonly string $username;
```

## Usage

Example of use for this library:

### Create a new instance using the current URL

```php
use Josantonius\Url\Url;

$url = new Url();
```

### Create a new instance using custom URL

```php
use Josantonius\Url\Url;

$url = new Url('https://domain.com');
```

### Gets authority

```php
use Josantonius\Url\Url;

$url = new Url(); // https://domain.com

$url->authority;  // "domain.com"


$url = new Url('https://user:pass@sub.domain.com:90/en/');

$url->authority; // "user:pass@sub.domain.com:90"


$url = new Url('https://user:pass@sub.domain.com/en/');

$url->authority; // "user:pass@sub.domain.com"


$url = new Url('https://sub.domain.com/en/');

$url->authority; // "sub.domain.com"
```

### Gets base URL

```php
use Josantonius\Url\Url;

$url = new Url(); // https://user:pass@domain.com:80/en/

$url->base; // "https://domain.com"


$url = new Url('https://domain.com:80/?tag=bug');

$url->base; // "https://domain.com:80"


$url = new Url('https://domain.com/en/');

$url->base; // "https://domain.com"
```

### Gets the path basename

```php
use Josantonius\Url\Url;

$url = new Url(); // https://domain.com/search.php

$url->basename; // "search.php"


$url = new Url('https://domain.com/en/web/docs/search.php?tag=bug');

$url->basename; // "search.php"


$url = new Url('https://domain.com/en/web/docs?tag=bug');

$url->basename; // "docs"
```

### Gets the path dirname

```php
use Josantonius\Url\Url;

$url = new Url(); // https://domain.com/search.php

$url->dirname; // "/"


$url = new Url('https://domain.com/en/web/docs/search.php?tag=bug');

$url->dirname; // "/en/web/docs"


$url = new Url('https://domain.com/en/web/docs?tag=bug');

$url->dirname; // "/en/web"
```

### Gets the path basename extension

```php
use Josantonius\Url\Url;

$url = new Url(); // https://domain.com/search.php

$url->extension; // "php"


$url = new Url('https://domain.com/en/web/docs/search.php?tag=bug');

$url->extension; // "php"


$url = new Url('https://domain.com/en/web/docs?tag=bug');

$url->extension; // ""
```

### Gets the path filename

```php
use Josantonius\Url\Url;

$url = new Url(); // https://domain.com/search.php

$url->filename; // "search"


$url = new Url('https://domain.com/en/web/docs/search.php?tag=bug');

$url->filename; // "search"


$url = new Url('https://domain.com/docs?tag=bug');

$url->filename; // "docs"
```

### Gets fragment

```php
use Josantonius\Url\Url;

$url = new Url(); // https://domain.com#top

$url->fragment; // "top"


$url = new Url('https://domain.com/en/web/docs#top');

$url->fragment; // "top"


$url = new Url('https://domain.com');

$url->fragment; // ""
```

### Gets the full URL

```php
use Josantonius\Url\Url;

$url = new Url(); // https://domain.com:80

$url->full;  // "https://domain.com"


$url = new Url('https://user:pass@sub.domain.com:90/en/');

$url->full; // "https://user:pass@sub.domain.com:90/en/"
```

### Gets hashed fragment

```php
use Josantonius\Url\Url;

$url = new Url(); // https://domain.com#top

$url->hash; // "#top"


$url = new Url('https://domain.com/en/web/docs#top');

$url->hash; // "#top"


$url = new Url('https://domain.com');

$url->hash; // ""
```

### Gets host

```php
use Josantonius\Url\Url;

$url = new Url(); // https://sub.domain.com

$url->host; // "sub.domain.com"


$url = new Url('https://sub.domain.com/en/web/docs#top');

$url->host; // "sub.domain.com"


$url = new Url('https://domain.com');

$url->host; // "domain.com"


$url = new Url('https://localhost');

$url->host; // "localhost"
```

### Gets path

```php
use Josantonius\Url\Url;

$url = new Url(); // https://domain.com/en

$url->path; // "/en/web/docs/search.php"


$url = new Url('https://domain.com/en/web/docs/search.php');

$url->path; // "/en/web/docs/search.php"


$url = new Url('https://domain.com/en/web/docs/');

$url->path; // "/en/web/docs/"


$url = new Url('https://domain.com/en?tag=bug');

$url->path; // "/en"


$url = new Url('https://domain.com?tag=bug');

$url->path; // ""
```

### Gets the query parameters

```php
use Josantonius\Url\Url;

$url = new Url(); // https://domain.com/en?tag=bug&order=asc#top

$url->parameters; // ["tag" => "bug", "order" => "asc"]


$url = new Url('https://domain.com/en/web/docs/search.php');

$url->parameters; // ""
```

### Gets password

```php
use Josantonius\Url\Url;

$url = new Url(); // https://:pass@domain.com

$url->password; // "pass"


$url = new Url('https://user:pass@domain.com');

$url->password; // "pass"


$url = new Url('https://user@domain.com');

$url->password; // ""
```

### Gets port

```php
use Josantonius\Url\Url;

$url = new Url(); // https://domain.com:90

$url->port; // 90


$url = new Url(); // https://domain.com:80

$url->port; // ""


$url = new Url(); // https://domain.com:443

$url->port; // ""


$url = new Url('https://domain.com:80/en/');

$url->port; // 80


$url = new Url('https://domain.com:443/en/');

$url->port; // 443


$url = new Url('https://domain.com/en/');

$url->port; // ""
```

### Gets scheme

```php
use Josantonius\Url\Url;

$url = new Url(); // http://domain.com

$url->scheme; // "http"


$url = new Url('https://domain.com');

$url->scheme; // "https"
```

### Gets path segments

```php
use Josantonius\Url\Url;

$url = new Url(); // https://domain.com?tag=bug

$url->segments; // []


$url = new Url('https://domain.com/en/web/docs/search.php');

$url->segments; // ['en', 'web', 'docs', 'search.php']
```

### Gets query

```php
use Josantonius\Url\Url;

$url = new Url(); // https://domain.com?tag=bug

$url->query; // "tag=bug"


$url = new Url('https://domain.com?tag=bug&order=asc#top');

$url->query; // "tag=bug&order=asc"

$url = new Url('https://domain.com');

$url->query; // ""
```

### Gets username

```php
use Josantonius\Url\Url;

$url = new Url(); // https://user@domain.com

$url->username; // "user"


$url = new Url('https://:pass@domain.com');

$url->username; // ""


$url = new Url('https://user:pass@domain.com');

$url->username; // "user"


$url = new Url('https://domain.com');

$url->username; // ""
```

## Tests

To run [tests](tests) you just need [composer](http://getcomposer.org/download/)
and to execute the following:

```console
git clone https://github.com/josantonius/php-url.git
```

```console
cd php-url
```

```console
composer install
```

Run unit tests with [PHPUnit](https://phpunit.de/):

```console
composer phpunit
```

Run code standard tests with [PHPCS](https://github.com/squizlabs/PHP_CodeSniffer):

```console
composer phpcs
```

Run [PHP Mess Detector](https://phpmd.org/) tests to detect inconsistencies in code style:

```console
composer phpmd
```

Run all previous tests:

```console
composer tests
```

## TODO

- [ ] Add new feature
- [ ] Improve tests
- [ ] Improve documentation
- [ ] Improve English translation in the README file
- [ ] Refactor code for disabled code style rules (see phpmd.xml and phpcs.xml)

## Changelog

Detailed changes for each release are documented in the
[release notes](https://github.com/josantonius/php-url/releases).

## Contribution

Please make sure to read the [Contributing Guide](.github/CONTRIBUTING.md), before making a pull
request, start a discussion or report a issue.

Thanks to all [contributors](https://github.com/josantonius/php-url/graphs/contributors)! :heart:

## Sponsor

If this project helps you to reduce your development time,
[you can sponsor me](https://github.com/josantonius#sponsor) to support my open source work :blush:

## License

This repository is licensed under the [MIT License](LICENSE).

Copyright © 2017-present, [Josantonius](https://github.com/josantonius#contact)


CHANGELOG.md
# CHANGELOG

## [v2.0.1](https://github.com/josantonius/php-url/releases/tag/v2.0.1) (2022-09-29)

* The notation type in the test function names has been changed from camel to snake case for readability.

* Functions were added to document the methods and avoid confusion.

* Disabled the ´CamelCaseMethodName´ rule in ´phpmd.xml´ to avoid warnings about function names in tests.

* The alignment of the asterisks in the comments has been fixed.

* Tests for Windows have been added.

* Tests for PHP 8.2 have been added.

## [v2.0.0](https://github.com/josantonius/php-url/releases/tag/v2.0.0) (2022-08-17)

> Version 1.x is considered as deprecated and unsupported.
> In this version (2.x) the library was completely restructured.
> It is recommended to review the documentation for this version and make the necessary changes
> before starting to use it, as it not be compatible with version 1.x.

---

* The library was completely refactored.

* Support for PHP version 8.1.

* Support for earlier versions of PHP **8.1** is discontinued.

* Improved documentation; `README.md`, `CODE_OF_CONDUCT.md`, `CONTRIBUTING.md` and `CHANGELOG.md`.

* Removed `Codacy`.

* Removed `PHP Coding Standards Fixer`.

* The `master` branch was renamed to `main`.

* The `develop` branch was added to use a workflow based on `Git Flow`.

* `Travis` is discontinued for continuous integration. `GitHub Actions` will be used from now on.

* Added `.github/CODE_OF_CONDUCT.md` file.
* Added `.github/CONTRIBUTING.md` file.
* Added `.github/FUNDING.yml` file.
* Added `.github/workflows/ci.yml` file.
* Added `.github/lang/es-ES/CODE_OF_CONDUCT.md` file.
* Added `.github/lang/es-ES/CONTRIBUTING.md` file.
* Added `.github/lang/es-ES/LICENSE` file.
* Added `.github/lang/es-ES/README` file.

* Deleted `.travis.yml` file.
* Deleted `.editorconfig` file.
* Deleted `CONDUCT.MD` file.
* Deleted `README-ES.MD` file.
* Deleted `.php_cs.dist` file.

## [1.2.1](https://github.com/josantonius/php-url/releases/tag/1.2.1) (2018-04-17)

* Bug fixed in `setUrlParams` method.

## [1.2.0](https://github.com/josantonius/php-url/releases/tag/1.2.0) (2018-02-22)

* Bugs fixed in `getBaseUrl` and `setUrlParams` method.

## [1.1.9](https://github.com/josantonius/php-url/releases/tag/1.1.9) (2018-01-07)

* The tests were fixed.

* Changes in documentation.

## [1.1.8](https://github.com/josantonius/php-url/releases/tag/1.1.8) (2017-11-12)

* Do more testings.
* Change the `addBackslash` method to `addBackSlash` method and refactoring.
* Fix command in composer.json will output the ERROR: The file "src,tests" does not exist.

* Added `Josantonius\Url\Tests\UrlTest::testGetProtocolWithUrl()` method.
* Added `Josantonius\Url\Tests\UrlTest::testGetDomainWithDomain()` method.
* Added `Josantonius\Url\Tests\UrlTest::testGetDomainWithInvalidDomain()` method.
* Added `Josantonius\Url\Tests\UrlTest::testAddBackSlashDefault()` method.

## [1.1.7](https://github.com/josantonius/php-url/releases/tag/1.1.7) (2017-11-09)

* Implemented `PHP Mess Detector` to detect inconsistencies in code styles.

* Implemented `PHP Code Beautifier and Fixer` to fixing errors automatically.

* Implemented `PHP Coding Standards Fixer` to organize PHP code automatically according to PSR standards.

## [1.1.6](https://github.com/josantonius/php-url/releases/tag/1.1.6) (2017-11-02)

* Implemented `PSR-4 autoloader standard` from all library files.

* Implemented `PSR-2 coding standard` from all library PHP files.

* Implemented `PHPCS` to ensure that PHP code complies with `PSR2` code standards.

* Implemented `Codacy` to automates code reviews and monitors code quality over time.

* Implemented `Codecov` to coverage reports.

* Added `Url/phpcs.ruleset.xml` file.

* Deleted `Url/src/bootstrap.php` file.

* Deleted `Url/tests/bootstrap.php` file.

* Deleted `Url/vendor` folder.

* Changed `Josantonius\Url\Test\UrlTest` class to  `Josantonius\Url\UrlTest` class.

## [1.1.5](https://github.com/josantonius/php-url/releases/tag/1.1.5) (2017-09-17)

* Unit tests supported by `PHPUnit` were added.

* The repository was synchronized with `Travis CI` to implement continuous integration.

* Added `Url/src/bootstrap.php` file

* Added `Url/tests/bootstrap.php` file.

* Added `Url/phpunit.xml.dist` file.
* Added `Url/_config.yml` file.
* Added `Url/.travis.yml` file.

* Added `Josantonius\Url\Url::setUrlParams()` method.

* Deleted `Josantonius\Url\Url::segment()` method.

* Added `Josantonius\Url\Url::segmentUri()` method.

* Deleted `Josantonius\Url\Tests\UrlTest` class.
* Deleted `Josantonius\Url\Tests\UrlTest::testGetCurrentPage()` method.
* Deleted `Josantonius\Url\Tests\UrlTest::testGetProtocol()` method.
* Deleted `Josantonius\Url\Tests\UrlTest::testIsSSL()` method.
* Deleted `Josantonius\Url\Tests\UrlTest::getDomain()` method.
* Deleted `Josantonius\Url\Tests\UrlTest::testGetUri()` method.
* Deleted `Josantonius\Url\Tests\UrlTest::testGetPort()` method.
* Deleted `Josantonius\Url\Tests\UrlTest::testAddBackslash()` method.
* Deleted `Josantonius\Url\Tests\UrlTest::testPrevious()` method.
* Deleted `Josantonius\Url\Tests\UrlTest::testRedirect()` method.
* Deleted `Josantonius\Url\Tests\UrlTest::testAutoLink()` method.
* Deleted `Josantonius\Url\Tests\UrlTest::testCustomAutoLink()` method.
* Deleted `Josantonius\Url\Tests\UrlTest::testGenerateSafeSlug()` method.
* Deleted `Josantonius\Url\Tests\UrlTest::testSegment()` method.
* Deleted `Josantonius\Url\Tests\UrlTest::testGetFirstSegment()` method.
* Deleted `Josantonius\Url\Tests\UrlTest::testGetLastSegment()` method.
* Deleted `Josantonius\Url\Tests\UrlTest::testGetBaseUrl()` method.

* Added `Josantonius\Url\Test\UrlTest` class.
* Added `Josantonius\Url\Test\UrlTest::testGetCurrentPage()` method.
* Added `Josantonius\Url\Test\UrlTest::testGetBaseUrl()` method.
* Added `Josantonius\Url\Test\UrlTest::testGetProtocol()` method.
* Added `Josantonius\Url\Test\UrlTest::testIsSSL()` method.

tests/UrlTest.php
<?php

/*
 * This file is part of https://github.com/josantonius/php-url repository.
 *
 * (c) Josantonius <hello@josantonius.dev>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * @phpcs:disable PSR1.Methods.CamelCapsMethodName.NotCamelCaps
 */

namespace Josantonius\Url\Tests;

use Josantonius\Url\Url;
use PHPUnit\Framework\TestCase;

class UrlTest extends TestCase
{
    public function test_should_set_the_current_url()
    {
        $currentUrl =
            'https://' .
            $_SERVER['SERVER_NAME'] . ':' .
            $_SERVER['SERVER_PORT'] .
            $_SERVER['REQUEST_URL'];

        $this->assertEquals($currentUrl, (new Url())->full);
    }

    /**
     * @runInSeparateProcess
     */
    public function test_should_not_contain_common_http_ports()
    {
        $_SERVER['SERVER_PORT'] = 80;

        $this->assertStringNotContainsString('80', (new Url())->full);

        $_SERVER['SERVER_PORT'] = 443;

        $this->assertStringNotContainsString('80', (new Url())->full);
    }

    /**
     * @runInSeparateProcess
     */
    public function test_should_contain_valid_scheme_when_is_secure_site()
    {
        $this->assertStringContainsString('https:', (new Url())->full);
    }

    /**
     * @runInSeparateProcess
     */
    public function test_should_contain_valid_scheme_when_is_non_secure_site()
    {
        $_SERVER['HTTPS'] = 'off';

        $this->assertStringContainsString('http:', (new Url())->full);
    }

    public function test_should_set_all_properties()
    {
        $url = new Url(
            'https://user:pass@sub.domain.com:90/en/web/docs/search.php?tag=bug&order=asc#top'
        );

        $this->assertEquals('https', $url->scheme);
        $this->assertEquals('user', $url->username);
        $this->assertEquals('pass', $url->password);
        $this->assertEquals('sub.domain.com', $url->host);
        $this->assertEquals(90, $url->port);
        $this->assertEquals('user:pass@sub.domain.com:90', $url->authority);
        $this->assertEquals('https://sub.domain.com:90', $url->base);
        $this->assertEquals('/en/web/docs/search.php', $url->path);
        $this->assertEquals('search.php', $url->basename);
        $this->assertEquals('/en/web/docs', $url->dirname);
        $this->assertEquals(["en", "web", "docs", "search.php"], $url->segments);
        $this->assertEquals('php', $url->extension);
        $this->assertEquals('search', $url->filename);
        $this->assertEquals('tag=bug&order=asc', $url->query);
        $this->assertEquals(["tag" => "bug", "order" => "asc"], $url->parameters);
        $this->assertEquals('top', $url->fragment);
        $this->assertEquals('#top', $url->hash);
    }

    public function test_should_set_nonexistent_properties_as_empty()
    {
        $url = new Url('');

        $this->assertEquals('', $url->scheme);
        $this->assertEquals('', $url->username);
        $this->assertEquals('', $url->password);
        $this->assertEquals('', $url->host);
        $this->assertEquals('', $url->port);
        $this->assertEquals('', $url->authority);
        $this->assertEquals('', $url->base);
        $this->assertEquals('', $url->path);
        $this->assertEquals('', $url->basename);
        $this->assertEquals('', $url->dirname);
        $this->assertEquals([], $url->segments);
        $this->assertEquals('', $url->extension);
        $this->assertEquals('', $url->filename);
        $this->assertEquals('', $url->query);
        $this->assertEquals([], $url->parameters);
        $this->assertEquals('', $url->fragment);
        $this->assertEquals('', $url->hash);
    }

    public function test_authority_component_should_be_set_up_without_the_port()
    {
        $url = new Url('https://user:pass@sub.domain.com');

        $this->assertEquals('user:pass@sub.domain.com', $url->authority);
    }

    public function test_authority_component_should_be_set_up_without_the_user_info()
    {
        $url = new Url('https://sub.domain.com');

        $this->assertEquals('sub.domain.com', $url->authority);
    }

    public function test_should_not_ignore_the_port_even_if_it_is_common_when_the_url_is_custom()
    {
        $url = new Url('https://user:pass@sub.domain.com:80');

        $this->assertEquals(80, $url->port);

        $url = new Url('https://user:pass@sub.domain.com:443');

        $this->assertEquals(443, $url->port);
    }
}


tests/bootstrap.php
<?php 
/**
 * Library for urls manipulation.
 * 
 * @author     Josantonius - hello@josantonius.com
 * @copyright  Copyright (c) 2017
 * @license    https://opensource.org/licenses/MIT - The MIT License (MIT)
 * @link       https://github.com/Josantonius/PHP-Url
 * @since      1.1.5
 */

$loader = require __DIR__ . '/../src/bootstrap.php';

$loader->add('Josantonius\Url\Test', __DIR__);

src/bootstrap.php
<?php 
/**
 * Library for urls manipulation.
 * 
 * @author     Josantonius - hello@josantonius.com
 * @copyright  Copyright (c) 2017
 * @license    https://opensource.org/licenses/MIT - The MIT License (MIT)
 * @link       https://github.com/Josantonius/PHP-Url
 * @since      1.1.5
 */

function includeIfExists($file) {

    if (file_exists($file)) {
        
        return include $file;
    }
}

if ((!$loader = includeIfExists(__DIR__ . '/../vendor/autoload.php')) && 
	(!$loader = includeIfExists(__DIR__ . '/../../../autoload.php'))) {
    
    die(PHP_EOL . 'You must set up the project dependencies, ' .
    	'run the following commands:' . PHP_EOL .
        PHP_EOL . 'curl -s http://getcomposer.org/installer | php' . PHP_EOL .
        PHP_EOL . 'php composer.phar install' . PHP_EOL . PHP_EOL);
}

return $loader;


src/Exception/UrlException.php
<?php 
/**
 * Library for urls manipulation.
 * 
 * @author     Josantonius - hello@josantonius.com
 * @copyright  Copyright (c) 2017
 * @license    https://opensource.org/licenses/MIT - The MIT License (MIT)
 * @link       https://github.com/Josantonius/PHP-Url
 * @since      1.0.0
 */

namespace Josantonius\Url\Exception;

/**
 * Exception class for Url library.
 *
 * You can use an exception and error handler with this library.
 *
 * @since 1.0.0
 *
 * @link https://github.com/Josantonius/PHP-ErrorHandler
 */
class UrlException extends \Exception { 

    /**
     * Exception handler.
     *
     * @since 1.0.0
     *
     * @param string $msg    → message error (Optional)
     * @param int    $error  → error code (Optional)
     * @param int    $status → HTTP response status code (Optional)
     */
    public function __construct($msg = '', $error = 0, $status = 0) {

        $this->message    = $msg;
        $this->code       = $error;
        $this->statusCode = $status;
    }
}


src/Url.php
<?php

declare(strict_types=1);

/*
 * This file is part of https://github.com/josantonius/php-url repository.
 *
 * (c) Josantonius <hello@josantonius.dev>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Josantonius\Url;

/**
 * Access URL components.
 */
class Url
{
    /**
     * The authority of the URL, in "[user-info@][host][:port]" format.
     *
     * @var string URL authority or empty string.
     */
    public readonly string $authority;

    /**
     * The base URL, in "[scheme:][//domain][:port]" format.
     *
     * @var string Base URL or empty string.
     */
    public readonly string $base;

    /**
     * The path basename of the URL, in "[filename][.extension]" format.
     *
     * @var string URL path basename or empty string.
     */
    public readonly string $basename;

    /**
     * The path dirname of the URL, in "[dirname]" format.
     *
     * @var string URL path dirname or empty string.
     */
    public readonly string $dirname;

    /**
     * The path basename extension of the URL, in "[extension]" format.
     *
     * @var string URL path basename extension or empty string.
     */
    public readonly string $extension;

    /**
     * The path filename of the URL, in "[filename]" format.
     *
     * @var string URL path filename or empty string.
     */
    public readonly string $filename;

    /**
     * The fragment of the URL, in "[fragment]" format.
     *
     * @var string URL fragment or empty string.
     */
    public readonly string $fragment;

    /**
     * The full URL.
     */
    public readonly string $full;

    /**
     * The hashed fragment of the URL, in "[#fragment]" format.
     *
     * @var string URL hashed fragment or empty string.
     */
    public readonly string $hash;

    /**
     * The host of the URL, in "[subdomain.][domain][.tld]" format.
     *
     * @var string URL host or empty string.
     */
    public readonly string $host;

    /**
     * The path of the URL, in "[path]" format.
     *
     * @var string URL path or empty string.
     */
    public readonly string $path;

    /**
     * The query parameters of the URL, in array format.
     *
     * @var array<string, mixed> URL query or empty string.
     */
    public readonly array $parameters;

    /**
     * The password of the URL, in "[password]" format.
     *
     * @var string URL password or empty string.
     */
    public readonly string $password;

    /**
     * The port of the URL, in "[port]" format.
     *
     * @var string URL port or empty string.
     */
    public readonly int|string $port;

    /**
     * The scheme of the URL, in "[scheme]" format.
     *
     * @var string URL scheme or empty string.
     */
    public readonly string $scheme;

    /**
     * The path segments of the URL, in array format.
     *
     * @var string[] URL path segments or empty array.
     */
    public readonly array $segments;

    /**
     * The query of the URL, in "[query]" format.
     *
     * @var string URL query or empty string.
     */
    public readonly string $query;

    /**
     * The user name of the URL, in "[username]" format.
     *
     * @var string URL username or empty string.
     */
    public readonly string $username;

    /**
     * Result of parse URL.
     */
    private array|bool $components;

    /**
     * Creates new instance.
     *
     * If no URL is provided, the URL of the current page will be generated.
     * The generated URL will not include ports 80 and 443 in it.
     */
    public function __construct(?string $url = null)
    {
        $this->full = $url ?? $this->getCurrentUrl();

        $components = parse_url($this->full) ?: [];

        $this->components = [...$components, ...pathinfo($components['path'] ?? '')];

        $this->setAttributes();

        unset($this->components);
    }

    /**
     * Sets URL scheme.
     */
    private function getCurrentUrl(): string
    {
        $path  = $_SERVER['REQUEST_URL'] ?? '';
        $port  = $_SERVER['SERVER_PORT'] ?? '';
        $host  = $_SERVER['SERVER_NAME'] ?? '';
        $https = $_SERVER['HTTPS']       ?? '';

        $scheme = $https === 'on' ? 'https:' : 'http:';

        $port = in_array($port, ['80', '443']) ? '' : ':' .  $port;

        return strtolower($scheme . ($host ? '//' . $host : '')) . $port . $path;
    }

    private function setAttributes(): void
    {
        $this->setFragment();
        $this->setHost();
        $this->setHash();
        $this->setPath();
        $this->setPort();
        $this->setSegments();
        $this->setQuery();
        $this->setParameters();
        $this->setFilename();
        $this->setDirname();
        $this->setBasename();
        $this->setExtension();
        $this->setScheme();
        $this->setUsername();
        $this->setPassword();
        $this->setAuthority();
        $this->setBase();
    }

    private function setAuthority(): void
    {
        $port     = $this->port     ? ':' . $this->port     : '';
        $password = $this->password ? ':' . $this->password : '';

        $userInfo = ($this->username . $password) ? $this->username . $password . '@' : '';

        $this->authority = $userInfo . $this->host . $port;
    }

    private function setBase(): void
    {
        $scheme = $this->scheme ? $this->scheme . ':' : '';
        $host   = $this->host   ? '//' . $this->host  : '';
        $port   = $this->port   ? ':' . $this->port   : '';

        $this->base = $scheme . $host . $port;
    }

    private function setBasename(): void
    {
        $this->basename = $this->components['basename'] ?? '';
    }

    private function setDirname(): void
    {
        $this->dirname = $this->components['dirname'] ?? '';
    }

    private function setExtension(): void
    {
        $this->extension = $this->components['extension'] ?? '';
    }

    private function setFilename(): void
    {
        $this->filename = $this->components['filename'] ?? '';
    }

    private function setFragment(): void
    {
        $this->fragment = $this->components['fragment'] ?? '';
    }

    private function setHash(): void
    {
        $this->hash = $this->fragment ? '#' . $this->fragment : '';
    }

    private function setHost(): void
    {
        $this->host = $this->components['host'] ?? '';
    }

    private function setParameters(): void
    {
        $params = [];

        parse_str($this->query, $params);

        $this->parameters = $params;
    }

    private function setPassword(): void
    {
        $this->password = $this->components['pass'] ?? '';
    }

    private function setPath(): void
    {
        $this->path = $this->components['path'] ?? '';
    }

    private function setPort(): void
    {
        $this->port = $this->components['port'] ?? '';
    }

    private function setQuery(): void
    {
        $this->query = $this->components['query'] ?? '';
    }

    private function setScheme(): void
    {
        $this->scheme = $this->components['scheme'] ?? '';
    }

    private function setSegments(): void
    {
        $this->segments = $this->path ? explode('/', trim($this->path, '/')) : [];
    }

    private function setUsername(): void
    {
        $this->username = $this->components['user'] ?? '';
    }
}
